### ISSUE TITLE
Support SCRAM-SHA-256 authentication

### ISSUE THREAD

**Issue by kravietz** [wal-g/wal-g#237]
When the server is using `SCRAM-SHA-256` (now officially recommended starting from PostgreSQL 10) `wal-g` throws the following error message:

```
INFO: 2019/04/08 15:39:38.782426 Doing full backup.
ERROR: 2019/04/08 15:39:38.798466 Connect: postgres connection failed: unknown authentication type: 10
```

**Comment by kravietz** [comment]
Ok, I now see it's an actually issue with the underlying library - if anyone else comes here from DuckDuckGo can you please upvote https://github.com/jackc/pgx/issues/451

**Comment by x4m** [comment]
pgx now supports SCRAM, I think we will update our pgx vendoring to v3.4.0 and this will be supported by WAL-G too.

**Comment by kravietz** [comment]
@x4m Any luck with this one?

**Comment by x4m** [comment]
@kravietz sorry, still had no time to test... can you please send a PR with bump of dependency here https://github.com/wal-g/wal-g/blob/master/Gopkg.lock#L258 and try if SCRAM is working?

**Comment by kravietz** [comment]
@x4m I'm not very familiar with Go and I can't even get your unmodified code to compile - it crashes with:

```
(cd main/pg && go build -tags "brotli lzo" -o wal-g -ldflags "-s -w -X github.com/wal-g/wal-g/cmd/pg.BuildDate=`date -u +%Y.%m.%d_%H:%M:%S` -X github.com/wal-g/wal-g/cmd/pg.GitRevision=`git rev-parse --short HEAD` -X github.com/wal-g/wal-g/cmd/pg.WalgVersion=`git tag -l --points-at HEAD`")
# github.com/wal-g/wal-g/vendor/github.com/cyberdelia/lzo
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/liblzo2.a(lzo1x_9x.o): relocation R_X86_64_32S against `.rodata' can not be used when making a PIE object; recompile with -fPIC
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/liblzo2.a(lzo_util.o): relocation R_X86_64_32 against `.rodata' can not be used when making a PIE object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
```

**Comment by x4m** [comment]
I think we should disable most of codecs like lzo and brotli in dev versions, because they create too much of hassle for new potential contributors....

I've updated pgx in related PR. I think if integration tests go well we can merge it.

**Comment by kravietz** [comment]
Doesn't seem to be really working :cry: using the latest release binary:

```
postgres@tyler:~$ /usr/bin/envdir /etc/wal-e.d /usr/local/bin/wal-g backup-push /var/lib/postgresql/12/main/
INFO: 2019/12/27 15:01:01.490527 Doing full backup.
ERROR: 2019/12/27 15:01:01.503507 unexpected message type
ERROR: 2019/12/27 15:01:01.503540 Failed to connect using provided PGHOST and PGPORT, trying localhost:5432
ERROR: 2019/12/27 15:01:01.523069 Connect: postgres connection failed: unexpected message type
```
`strace` suggests it happens specifically during the authentication stage:

```
write(5, "\0\0\0\27\0\3\0\0user\0postgres\0\0", 23) = 23
read(5, 0xc0003ac000, 8192)             = -1 EAGAIN (Resource temporarily unavailable)
epoll_pwait(4, [], 128, 0, NULL, 0)     = 0
epoll_pwait(4, [{EPOLLIN|EPOLLOUT, {u32=1907695152, u64=140683561475632}}], 128, -1, NULL, 0) = 1
read(5, "R\0\0\0\27\0\0\0\nSCRAM-SHA-256\0\0", 8192) = 24
getrandom("\x56\xb7\xd0\x5c\xf7\x1b\x29\x12\xbe\xca\xe4\xf3\x7f\xc2\xb2\x2e\x51\x1d", 18, 0) = 18
write(5, "p\0\0\0006SCRAM-SHA-256\0\0\0\0 n,,n=,r=V"..., 55) = 55
read(5, "R\0\0\0\\\0\0\0\vr=VrfQXPcbKRK+yuTzf8KyL"..., 8168) = 93
write(5, "p\0\0\0lc=biws,r=VrfQXPcbKRK+yuTzf8"..., 109) = 109
read(5, "E\0\0\0hSFATAL\0VFATAL\0C28P01\0Mpassw"..., 8075) = 105
```

**Comment by kravietz** [comment]
@x4m It seems like SCRAM only appeared in 4.6.0 of pgx https://github.com/jackc/pgx/commit/5044e8473ad948114b6cb63f6f30f94fc7834667 and you bumped to 3.6.0.

**Comment by kravietz** [comment]
@x4m I was wrong, it also appears here https://github.com/jackc/pgx/blob/v3.6.0/auth_scram.go No idea why it's failing then.

**Comment by Neustradamus** [comment]
@kravietz, @x4m, @sebasmannem, @kjmph: SCRAM-SHA-256 is supported now?

This ticket can be closed?

**Comment by x4m** [comment]
I can’t verify right now. If anyone can test and confirm that scram works - let’s close the issue.

**Comment by kjmph** [comment]
I’m not in front of that environment right now, but it does work with PGX v3, as long as the PGPASSWORD env var is used. I was running it for quite a while.

**Comment by kjmph** [comment]
Yes, that's correct. Which to be fair, PGCONN has moved forward a lot in this area. New versions of PGX support PGCONN/PGPASSFILE Configs:

https://github.com/jackc/pgx/commit/b3c8a73dc7e9385888fafbd08124e8f10c69261f

PGPASSFILE is its own package now:

https://github.com/jackc/pgpassfile

So, creating a Config object and passing it along would allow WAL-G's config to be used.

**Comment by kjmph** [comment]
Does WAL-G support a config file? If so, it could pass through configuration to PG* modules via the Config parameter to Connect. That, I think, is @OrangeDog's request. Probably belongs in its own issue. However, from my end everything works great, I am able to use the environment variables currently supported. I would request the documentation to stop claiming PGPASSFILE works, as it doesn't: https://wal-g.readthedocs.io/PostgreSQL/#configuration To support PGPASSFILE, WAL-G would have to upgrade dependencies, which is probably a heavy lift for the team.

**Comment by x4m** [comment]
WAL-G passes parameters from config file to env vars automatically. For pgx and for storages. So this should work. AFAIK it passes PGPASSFILE too.

**Comment by x4m** [comment]
We do not depend on features of pgx much. I think it should be straightforward to upgrade, more or less…

**Comment by kjmph** [comment]
Whoa, this is buried in Greenplum documentation? https://wal-g.readthedocs.io/Greenplum/#configuration-file

I never knew there was a config file. I've been passing nothing but ENV vars. Then, I propose @OrangeDog open a new ticket if WAL-G converts the config file to ENV vars automatically.

Also, no, PGPASSFILE isn't supported because WAL-G uses PGX 3.6.0, which only processes ENV vars and PGPASSFILE isn't a supported ENV var for this version. Maybe you mean WAL-G converts PGPASSFILE from the config file to ENV vars? The dependency chain doesn't use this ENV var, so that seems an issue:

https://github.com/jackc/pgx/blob/c73e7d75061bb42b0282945710f344cfe1113d10/conn.go#L1263

This is turning into a fascinating ride.. If y'all find it straightforward to upgrade that would be interesting to see. I tried in a local copy and immediately ran into headaches.
